### PR TITLE
Virtual media resource migration

### PR DESCRIPTION
**Pull request by prajwalpatil25**
Migrated virtual media resource.
image

### CODE REVIEW

**Comment by rounak-adhikary** [comment]
Please add the coverage report

File: redfish/provider/resource_redfish_virtual_media.go [hunk 1]
```diff
@@ -1,26 +1,456 @@
...
+		return
+	}
+	// Validate image extension
+	image := plan.Image.ValueString()
+	if !strings.HasSuffix(image, ".iso") && !strings.HasSuffix(image, ".img") {
+		resp.Diagnostics.AddError(mountError, "Unable to Process the request. Image extension should be .iso or .img")
+		return
+	}
+	// Validate transfer method
+	if plan.TransferMethod.ValueString() == "Upload" {
```
Comment (rounak-adhikary) [inline redfish/provider/resource_redfish_virtual_media.go:145]: If we are not supporting Upload as an option, why not just have this logic as part of the validator?
We can have something like
```
stringvalidator.OneOf([]string{
	"Stream",
	// "Upload", # TODO
}...),
```

File: redfish/provider/resource_redfish_virtual_media.go [hunk 2]
```diff
@@ -1,26 +1,456 @@
...
+			Description:         "Indicates how the data is transferred",
+			MarkdownDescription: "Indicates how the data is transferred",
+			Validators: []validator.String{
+				stringvalidator.OneOf([]string{
+					"Stream",
+					"Upload",
+				}...),
+			},
+		},
+		"transfer_protocol_type": schema.StringAttribute{
```
Comment (rounak-adhikary) [inline redfish/provider/resource_redfish_virtual_media.go:90]: `transfer_protocol_type` and `transfer_method` are Optional and Computed. Are there default values?
I see that we are using code like 
```
virtualMediaConfig := redfish.VirtualMediaConfig{
		Image:                image,
		Inserted:             plan.Inserted.ValueBool(),
		TransferMethod:       plan.TransferMethod.ValueString(),
		TransferProtocolType: plan.TransferProtocolType.ValueString(),
		WriteProtected:       plan.WriteProtected.ValueBool(),
	}
```
Please check if these will work correctly without default values.

File: redfish/provider/resource_redfish_virtual_media.go [hunk 3]
```diff
@@ -1,26 +1,456 @@
...
+	virtualMediaCollection, err := systems[0].VirtualMedia()
+	if err != nil {
+		resp.Diagnostics.AddError("Couldn't retrieve virtual media collection from redfish API", err.Error())
+		return
+	}
+	if len(virtualMediaCollection) != 0 {
+		for index := range virtualMediaCollection {
+			virtualMedia, err := insertMedia(virtualMediaCollection[index].ID, virtualMediaCollection, virtualMediaConfig, service)
+			if err != nil {
+				resp.Diagnostics.AddError("Error: %s", err.Error())
```
Comment (rounak-adhikary) [inline redfish/provider/resource_redfish_virtual_media.go:182]: String formatting won't work here. Suggest something like
```suggestion
				resp.Diagnostics.AddError("Error while inserting virtual media", err.Error())
```

File: redfish/provider/resource_redfish_virtual_media.go [hunk 4]
```diff
@@ -1,26 +1,456 @@
...
+		var virtualMediaID string
+		if strings.HasSuffix(plan.Image.ValueString(), ".iso") {
+			virtualMediaID = "CD"
+		} else {
+			virtualMediaID = "RemovableDisk"
+		}
+
+		virtualMedia, err := insertMedia(virtualMediaID, virtualMediaCollection, virtualMediaConfig, service)
+		if err != nil {
+			resp.Diagnostics.AddError("Error: %s", err.Error())
```
Comment (rounak-adhikary) [inline redfish/provider/resource_redfish_virtual_media.go:218]: Same here

File: redfish/provider/resource_redfish_virtual_media.go (hunk 1 above)
Comment (prajwalpatil25) [reply redfish/provider/resource_redfish_virtual_media.go:145]: As this limitation is from iDRAC, we explicitly want to mention that it is not supported by the implementation.

File: redfish/provider/resource_redfish_virtual_media.go (hunk 2 above)
Comment (prajwalpatil25) [reply redfish/provider/resource_redfish_virtual_media.go:90]: tested, as this are optional plus computed it shows them as "known after apply" and creates the resource successfully.

  ```
+ resource "redfish_virtual_media" "vm" {
      + id                     = (known after apply)
      + image                  = "http://linuxlib.us.dell.com/pub/redhat/RHEL8/8.8/BaseOS/x86_64/os/images/efiboot.img"
      + inserted               = (known after apply)
      + redfish_server         = {
          + endpoint = "https://10.226.197.128"
          + password = (sensitive value)
          + user     = "root"
        }
      + transfer_method        = (known after apply)
      + transfer_protocol_type = (known after apply)
      + write_protected        = true
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

redfish_virtual_media.vm["my-server-1"]: Creating...
redfish_virtual_media.vm["my-server-1"]: Creation complete after 6s [id=/redfish/v1/Managers/iDRAC.Embedded.1/VirtualMedia/RemovableDisk]

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.

```

File: redfish/provider/resource_redfish_virtual_media.go (hunk 3 above)
Comment (prajwalpatil25) [reply redfish/provider/resource_redfish_virtual_media.go]: done

File: redfish/provider/resource_redfish_virtual_media.go (hunk 4 above)
Comment (prajwalpatil25) [reply redfish/provider/resource_redfish_virtual_media.go]: done

File: redfish/provider/resource_redfish_virtual_media.go (hunk 2 above)
Comment (rounak-adhikary) [reply redfish/provider/resource_redfish_virtual_media.go:90]: then fine

File: redfish/provider/resource_redfish_virtual_media.go (hunk 1 above)
Comment (rounak-adhikary) [reply redfish/provider/resource_redfish_virtual_media.go:145]: ok

**Comment by rounak-adhikary** [comment]
One last thing, we also need to generate the documentation.

**Review by rounak-adhikary — APPROVED** [review]
lgtm

**Review by Krishnan-Priyanshu — DISMISSED** [review]
LGTM

File: redfish/provider/resource_redfish_virtual_media.go [hunk 5]
```diff
@@ -1,26 +1,456 @@
...
+	return &virtualMediaResource{}
+}
+
+// virtualMediaResource is the resource implementation.
+type virtualMediaResource struct {
+	p *redfishProvider
+}
+
+const (
+	serviceError = "service error"
```
Comment (Sakshi-dell) [inline redfish/provider/resource_redfish_virtual_media.go:37]: There's a new file constants.go where these common error messages are present, you can re-use the same and add new error messages there only

Comment (prajwalpatil25) [reply redfish/provider/resource_redfish_virtual_media.go]: done

**Review by Sakshi-dell — APPROVED** [review]

**Review by Krishnan-Priyanshu — APPROVED** [review]

**Pull request merged**